From a59e090f82c2cb9884d83687e2b93ed698c6f5ca Mon Sep 17 00:00:00 2001
From: Ferry Toth <ftoth@exalondelft.nl>
Date: Sat, 26 Apr 2025 17:43:17 +0200
Subject: [PATCH 1/1] glibc: turn off _GNU_SOURCE symbols

glibc 2.38 would include  __isoc23_strtol and similar symbols. This is triggerd by
_GNU_SOURCE but we have to set that for other definitions. Therefore play with defines
to turn this off within pseudo_wrappers.c. Elsewhere we can switch to _DEFAULT_SOURCE
rather than _GNU_SOURCE.

Upstream-Status: Pending
Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>
---
 pseudo_client.c   |  2 +-
 pseudo_util.c     | 11 +++++++++++
 pseudo_wrappers.c | 12 ++++++++++++
 3 files changed, 24 insertions(+), 1 deletion(-)

diff --git a/pseudo_client.c b/pseudo_client.c
index a03d6b1..5efc239 100644
--- a/pseudo_client.c
+++ b/pseudo_client.c
@@ -6,7 +6,7 @@
  * SPDX-License-Identifier: LGPL-2.1-only
  *
  */
-#define _GNU_SOURCE
+#define _DEFAULT_SOURCE
 
 #include <stdio.h>
 #include <signal.h>
diff --git a/pseudo_util.c b/pseudo_util.c
index 24115ff..93dccdd 100644
--- a/pseudo_util.c
+++ b/pseudo_util.c
@@ -8,6 +8,17 @@
  */
 /* we need access to RTLD_NEXT for a horrible workaround */
 #define _GNU_SOURCE
+/* glibc 2.38 would include  __isoc23_strtol and similar symbols. This is trggerd by
+ * _GNU_SOURCE but we have to set that for other definitions. Therefore play with defines
+ * to turn this off.
+ */
+#include <features.h>
+#undef __GLIBC_USE_ISOC2X
+#undef __GLIBC_USE_C2X_STRTOL
+#define __GLIBC_USE_C2X_STRTOL 0
+#undef __GLIBC_USE_ISOC23
+#undef __GLIBC_USE_C23_STRTOL
+#define __GLIBC_USE_C23_STRTOL 0
 
 #include <ctype.h>
 #include <errno.h>
diff --git a/pseudo_wrappers.c b/pseudo_wrappers.c
index 9ae1200..96039c0 100644
--- a/pseudo_wrappers.c
+++ b/pseudo_wrappers.c
@@ -6,6 +6,18 @@
  * SPDX-License-Identifier: LGPL-2.1-only
  *
  */
+/* glibc 2.38 would include  __isoc23_strtol and similar symbols. This is trggerd by
+ * _GNU_SOURCE but we have to set that for other definitions. Therefore play with defines
+ * to turn this off.
+ */
+#include <features.h>
+#undef __GLIBC_USE_ISOC2X
+#undef __GLIBC_USE_C2X_STRTOL
+#define __GLIBC_USE_C2X_STRTOL 0
+#undef __GLIBC_USE_ISOC23
+#undef __GLIBC_USE_C23_STRTOL
+#define __GLIBC_USE_C23_STRTOL 0
+
 #include <assert.h>
 #include <stdlib.h>
 #include <limits.h>
-- 
2.43.0

